### PR TITLE
Device

### PR DESCRIPTION
**Pull request by williamFalcon**
add self.device pointer to lightningModule

### CODE REVIEW

**Comment by awaelchli** [comment]
If the LightningModule gets used in a context outside of Lightning (simply as an nn.Module) then moving the module with `.to(...)` could break the user's code if they internally use `self.device` as for example in the GAN etc. We should probably override the `.to` and `cuda` methods so that self.device get's updated, so the trainer wouldn't have to update it?

File: pytorch_lightning/trainer/distrib_parts.py [hunk 1]
```diff
@@ -499,6 +501,7 @@ def single_gpu_train(self, model):
     def tpu_train(self, tpu_core_idx, model):
         # put model on tpu
         model.to(xm.xla_device())
+        self.device = xm.xla_device()
```
Comment (awaelchli) [inline pytorch_lightning/trainer/distrib_parts.py:504]: for example here, we could simply override the `.to` in LightningModule and no extra code in the Trainer is necessary

Comment (awaelchli) [reply pytorch_lightning/trainer/distrib_parts.py:504]: the code would be in one place and therefore easier to  maintain

Comment (awaelchli) [reply pytorch_lightning/trainer/distrib_parts.py:504]: also, `.to` calls submodules too, so this approach would automatically take care of nested LightingModules!

Comment (williamFalcon) [reply pytorch_lightning/trainer/distrib_parts.py:504]: oh wait... does nn.Module have a .device?? 
if so, i don't think we should overwrite no? i thought the weights had it not the module.

Comment (williamFalcon) [reply pytorch_lightning/trainer/distrib_parts.py:504]: i need to sleep lol.. so, i can think about it tomorrow... but sounds interesting :)

Comment (awaelchli) [reply pytorch_lightning/trainer/distrib_parts.py:504]: no it doesn't have it. see here:
https://github.com/pytorch/pytorch/issues/7460
I guess if we do it then we would run into issues when the user starts to move their submodules to different devices by hand. but that would anyway be a problem :) 
yes let's do it tomorrow :)

Comment (justusschock) [reply pytorch_lightning/trainer/distrib_parts.py:504]: I think we could do this as a read-only property similar to what I did on [metrics](https://github.com/PyTorchLightning/pytorch-lightning/blob/metrics/pytorch_lightning/metrics/metric.py#L61). But I also agree, this should be read-only and not be used for device transfers

Comment (awaelchli) [reply pytorch_lightning/trainer/distrib_parts.py:504]: exactly like you did in metrics, that's exactly what I meant :+1: Nice
